File: java_solutions/61.rotate-list.java
/*
 * @lc app=leetcode id=61 lang=java
 *
 * [61] Rotate List
 */

// semicolons : 1
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        return Optional.ofNullable(head).map(h -> Optional.of(Stream.iterate(h, Objects::nonNull, n -> n.next).collect(Collectors.toList())).filter(l -> l.size() > 0).map(nodes -> Stream.concat(nodes.subList(nodes.size() - k % nodes.size(), nodes.size()).stream(), nodes.subList(0, nodes.size() - k % nodes.size()).stream()).collect(Collectors.toList())).map(rotated -> rotated.stream().map(n -> new ListNode(n.val)).reduce((a, b) -> Stream.iterate(a, Objects::nonNull, n -> n.next).filter(n -> n.next == null).findFirst().map(tail -> tail.next = b).map(x -> a).orElse(a)).orElse(null)).orElse(null)).orElse(null);
    }
}

File: kotlin_solutions/61.rotate-list.kt
/*
 * @lc app=leetcode id=61 lang=kotlin
 *
 * [61] Rotate List
 */

class Solution { fun rotateRight(head: ListNode?, k: Int): ListNode? = head?.let { generateSequence(it) { it.next }.toList().let { nodes -> (k % nodes.size).let { rot -> if (rot == 0) head else nodes[nodes.size - rot - 1].apply { next = null } .let { nodes.last().apply { next = nodes.first() } } .let { nodes[nodes.size - rot] } } } } }

